### ISSUE TITLE
Defer PluginResolver

### ISSUE THREAD

**Issue by lukasniemeier-zalando** [zalando-incubator/put-it-to-rest#64]
**As** the `RestClientPostProcessor`
**I want** to defer usage of the `PluginResolver`
**So that** I do not violate the container by causing premature bean instantiation.

Currently the `RestClientPostProcessor` [depends](https://github.com/zalando-incubator/put-it-to-rest/blob/master/src/main/java/org/zalando/putittorest/RestClientAutoConfiguration.java#L26) on the `PluginResolver`, which seems to cause premature [bean instantiation](https://github.com/zalando-incubator/put-it-to-rest/blob/master/src/main/java/org/zalando/putittorest/DefaultPluginResolver.java#L30). [Documentation](http://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/beans/factory/config/BeanFactoryPostProcessor.html) states that a `BeanFactoryPostProcessor` should not do that.

I guess one of the effects of this is shown in the following log-line on startup:

```
Bean 'myBean' of type [class MyType] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
```

Can we defer resolving of the plugins?

**Comment by whiskeysierra** [comment]
Instead of looking up beans early we could write a custom plugin adapter that takes the type (class literal) and a (fallback) supplier (ie the constructor). It will wait for the first invocation in order to lookup the real plugin from the context. If it finds it, it will use it. If not we invoke the supplier and use that one. We basically take the existing code and move it to it's own class in order to delay everything.

**Comment by whiskeysierra** [comment]
@lukasniemeier-zalando Released as 0.12.1
